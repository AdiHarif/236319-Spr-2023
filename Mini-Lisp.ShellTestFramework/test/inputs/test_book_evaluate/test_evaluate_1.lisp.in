; =================================================
; Evaluate - lookup for atom value from the a-list:
; =================================================
(evaluate 'a '(a.T t.t nil.nil)) ; T
(evaluate 'a '(a.x t.t nil.nil)) ; X

; =================================================
; Evaluate - evaluation of 1 arg atomic functions:
; =================================================
(evaluate '(atom a) '(a.x t.t nil.nil))      ; T
(evaluate '(atom a) '(a.(x) t.t nil.nil))    ; NIL
(evaluate '(car a) '(a.(x) t.t nil.nil))     ; X
(evaluate '(car a) '(a.(x y z) t.t nil.nil)) ; X
(evaluate '(car a) '(a.(x.y) t.t nil.nil))   ; X.Y
(evaluate '(cdr a) '(a.(x) t.t nil.nil))     ; NIL
(evaluate '(cdr a) '(a.(x y z) t.t nil.nil)) ; (Y Z)
(evaluate '(cdr a) '(a.(x.y) t.t nil.nil))   ; NIL
(evaluate '(eval a) '(a.x x.y t.t nil.nil))  ; Y

; =================================================
; Evaluate - evaluation of 2 arg atomic functions:
; =================================================
(evaluate '(cons a b) '(a.x b.y t.t nil.nil))  ; (X.Y)
(evaluate '(cons a b) '(a.x b.x t.t nil.nil))  ; (X.X)
(evaluate '(eq a a) '(a.x t.t nil.nil))        ; T
(evaluate '(eq a b) '(a.x b.y t.t nil.nil))    ; NIL
(evaluate '(eq a b) '(a.x b.x t.t nil.nil))    ; T
(evaluate '(set a b) '(a.x b.y t.t nil.nil))   ; Y

; =================================================
; Evaluate - cond:
; =================================================
(evaluate '(cond (())) '(t.t nil.nil))                    ; NIL
(evaluate '(cond (t nil)) '(t.t nil.nil))                 ; NIL
(evaluate '(cond (a t)) '(a.t t.t nil.nil))               ; T
(evaluate '(cond (a b)) '(a.t b.t t.t nil.nil))           ; T
(evaluate '(cond (t a) (nil b)) '(a.a b.b t.t nil.nil))   ; A
(evaluate '(cond (nil a) (t b)) '(a.a b.b t.t nil.nil))   ; B
(evaluate '(cond (a b) (b a)) '(a.t b.t t.t nil.nil))     ; T
(evaluate '(cond (a b) (b a)) '(a.nil b.t t.t nil.nil))   ; NIL
(evaluate '(cond (a b) (a b) (b a)) '(a.nil b.t t.t nil.nil)) ; NIL

; =================================================
; Evaluate - the error function:
; =================================================
(evaluate '(error) '(t.t nil.nil)) ; NIL
(evaluate '(error a) '(a.my_err t.t nil.nil)) ; MY_ERR

; =================================================
; Evaluate - Errors:
; =================================================
(evaluate '(bla) '(t.t nil.nil))
(evaluate '(NIL NIL) '(t.t nil.nil))
