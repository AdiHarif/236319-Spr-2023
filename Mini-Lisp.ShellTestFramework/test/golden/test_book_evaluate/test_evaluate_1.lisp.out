T
X
T
NIL
X
X
(X.Y)
NIL
(Y Z)
NIL
Y
(X.Y)
(X.X)
T
NIL
T
Y
NIL
NIL
T
T
A
B
T
NIL
NIL
Traceback (most recent call last):
	From function EVALUATE:
		Args: ((ERROR) ((T.T) (NIL)))
	From function COND:
		Args: (((ATOM S-EXPRESSION) (LOOKUP S-EXPRESSION A-LIST)) ((IS-ATOMIC (CAR S-EXPRESSION)) (EVALUATE-ATOMIC S-EXPRESSION A-LIST)) (T (APPLY (EVALUATE (CAR S-EXPRESSION) A-LIST) (CDR S-EXPRESSION) A-LIST)))
	From function EVALUATE-ATOMIC:
		Args: ((ERROR) ((T.T) (NIL)))
	From function APPLY-ATOMIC:
		Args: (ERROR NIL ((T.T) (NIL)))
	From function COND:
		Args: (((EQ (QUOTE COND) ATOMIC) (EVALUATE-COND ACTUALS A-LIST)) (T (APPLY-EAGER-ATOMIC ATOMIC (EVALUATE-LIST ACTUALS A-LIST) A-LIST)))
	From function APPLY-EAGER-ATOMIC:
		Args: (ERROR NIL ((T.T) (NIL)))
	From function COND:
		Args: (((EQ ATOMIC (QUOTE ERROR)) (ERROR ACTUALS)) ((EQ ATOMIC (QUOTE EVAL)) (EVALUATE (CAR ACTUALS) A-LIST)) ((NULL (CDR ACTUALS)) (APPLY-TRIVIAL-ATOMIC ATOMIC (CAR ACTUALS) NIL)) (T (APPLY-TRIVIAL-ATOMIC ATOMIC (CAR ACTUALS) (CAR (CDR ACTUALS)))))
	From function ERROR:
		Args: NIL
Error [NIL] in (ERROR ACTUALS)
Traceback (most recent call last):
	From function EVALUATE:
		Args: ((ERROR A) ((A.MY_ERR) (T.T) (NIL)))
	From function COND:
		Args: (((ATOM S-EXPRESSION) (LOOKUP S-EXPRESSION A-LIST)) ((IS-ATOMIC (CAR S-EXPRESSION)) (EVALUATE-ATOMIC S-EXPRESSION A-LIST)) (T (APPLY (EVALUATE (CAR S-EXPRESSION) A-LIST) (CDR S-EXPRESSION) A-LIST)))
	From function EVALUATE-ATOMIC:
		Args: ((ERROR A) ((A.MY_ERR) (T.T) (NIL)))
	From function APPLY-ATOMIC:
		Args: (ERROR (A) ((A.MY_ERR) (T.T) (NIL)))
	From function COND:
		Args: (((EQ (QUOTE COND) ATOMIC) (EVALUATE-COND ACTUALS A-LIST)) (T (APPLY-EAGER-ATOMIC ATOMIC (EVALUATE-LIST ACTUALS A-LIST) A-LIST)))
	From function APPLY-EAGER-ATOMIC:
		Args: (ERROR (MY_ERR) ((A.MY_ERR) (T.T) (NIL)))
	From function COND:
		Args: (((EQ ATOMIC (QUOTE ERROR)) (ERROR ACTUALS)) ((EQ ATOMIC (QUOTE EVAL)) (EVALUATE (CAR ACTUALS) A-LIST)) ((NULL (CDR ACTUALS)) (APPLY-TRIVIAL-ATOMIC ATOMIC (CAR ACTUALS) NIL)) (T (APPLY-TRIVIAL-ATOMIC ATOMIC (CAR ACTUALS) (CAR (CDR ACTUALS)))))
	From function ERROR:
		Args: (MY_ERR)
Error [(MY_ERR)] in (ERROR ACTUALS)
Traceback (most recent call last):
	From function EVALUATE:
		Args: ((BLA) ((T.T) (NIL)))
	From function COND:
		Args: (((ATOM S-EXPRESSION) (LOOKUP S-EXPRESSION A-LIST)) ((IS-ATOMIC (CAR S-EXPRESSION)) (EVALUATE-ATOMIC S-EXPRESSION A-LIST)) (T (APPLY (EVALUATE (CAR S-EXPRESSION) A-LIST) (CDR S-EXPRESSION) A-LIST)))
	From function APPLY:
		Args: ((EVALUATE (CAR S-EXPRESSION) A-LIST) (CDR S-EXPRESSION) A-LIST)
	From function EVALUATE:
		Args: (BLA ((T.T) (NIL)))
	From function COND:
		Args: (((ATOM S-EXPRESSION) (LOOKUP S-EXPRESSION A-LIST)) ((IS-ATOMIC (CAR S-EXPRESSION)) (EVALUATE-ATOMIC S-EXPRESSION A-LIST)) (T (APPLY (EVALUATE (CAR S-EXPRESSION) A-LIST) (CDR S-EXPRESSION) A-LIST)))
	From function LOOKUP:
		Args: (BLA ((T.T) (NIL)))
	From function COND:
		Args: (((NULL A-LIST) (ERROR UNDEFINED ID)) ((EQ ID (CAR (CAR A-LIST))) (CDR (CAR A-LIST))) (T (LOOKUP ID (CDR A-LIST))))
	From function LOOKUP:
		Args: (BLA ((NIL)))
	From function COND:
		Args: (((NULL A-LIST) (ERROR UNDEFINED ID)) ((EQ ID (CAR (CAR A-LIST))) (CDR (CAR A-LIST))) (T (LOOKUP ID (CDR A-LIST))))
	From function LOOKUP:
		Args: (BLA NIL)
	From function COND:
		Args: (((NULL A-LIST) (ERROR UNDEFINED ID)) ((EQ ID (CAR (CAR A-LIST))) (CDR (CAR A-LIST))) (T (LOOKUP ID (CDR A-LIST))))
	From function ERROR:
		Args: UNDEFINED
Error [UNDEFINED] in UNDEFINED
Traceback (most recent call last):
	From function EVALUATE:
		Args: ((NIL NIL) ((T.T) (NIL)))
	From function COND:
		Args: (((ATOM S-EXPRESSION) (LOOKUP S-EXPRESSION A-LIST)) ((IS-ATOMIC (CAR S-EXPRESSION)) (EVALUATE-ATOMIC S-EXPRESSION A-LIST)) (T (APPLY (EVALUATE (CAR S-EXPRESSION) A-LIST) (CDR S-EXPRESSION) A-LIST)))
	From function APPLY:
		Args: (NIL (NIL) ((T.T) (NIL)))
	From function APPLY-DECOMPOSED-LAMBDA:
		Args: ((CAR LAMBDA-EXPRESSION) (CAR (CDR LAMBDA-EXPRESSION)) (CAR (CDR (CDR LAMBDA-EXPRESSION))) ACTUALS A-LIST)
	From function CAR:
		Args: NIL
Error [CAR] in NIL
